Devirtualize Event::eventInterface

rniwa opened the pull request:
#### bd2795436d8c4c389d30444f2e927919e2520cb8
<pre>
Devirtualize Event::eventInterface
<a href="https://bugs.webkit.org/show_bug.cgi?id=269338">https://bugs.webkit.org/show_bug.cgi?id=269338</a>

Reviewed by Chris Dumez and Yusuke Suzuki.

This PR replaces Event::eventInterface, which is a virtual function, with Event::interfaceType,
which is an inline non-virtual function that reads the value off of m_eventInterface set forth
in the Event constructor. WebKitAdditions&apos; files need to be migrated once this PR lands so they
currently fallback to calling Event::eventInterface (m_eventInterface is set to 0 in this case).

This PR also replaces EventInterface enum with EventInterfaceType enum class. Because one of
the valid values of EventInterface enum is EventInterfaceType (for Event class), we need to
explicitly write `enum EventInterfaceType` for now. The plan is to remove EventInterface once
the code in WebKitAdditions has migrated to use the new code path.

* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.cpp:
(WebCore::GPUUncapturedErrorEvent::GPUUncapturedErrorEvent):
* Source/WebCore/Modules/airplay/WebKitPlaybackTargetAvailabilityEvent.cpp:
(WebCore::WebKitPlaybackTargetAvailabilityEvent::WebKitPlaybackTargetAvailabilityEvent):
* Source/WebCore/Modules/applepay/ApplePayCancelEvent.cpp:
(WebCore::ApplePayCancelEvent::ApplePayCancelEvent):
* Source/WebCore/Modules/applepay/ApplePayCouponCodeChangedEvent.cpp:
(WebCore::ApplePayCouponCodeChangedEvent::ApplePayCouponCodeChangedEvent):
* Source/WebCore/Modules/applepay/ApplePayPaymentAuthorizedEvent.cpp:
(WebCore::ApplePayPaymentAuthorizedEvent::ApplePayPaymentAuthorizedEvent):
* Source/WebCore/Modules/applepay/ApplePayPaymentMethodSelectedEvent.cpp:
(WebCore::ApplePayPaymentMethodSelectedEvent::ApplePayPaymentMethodSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.cpp:
(WebCore::ApplePayShippingContactSelectedEvent::ApplePayShippingContactSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp:
(WebCore::ApplePayShippingMethodSelectedEvent::ApplePayShippingMethodSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayValidateMerchantEvent.cpp:
(WebCore::ApplePayValidateMerchantEvent::ApplePayValidateMerchantEvent):
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
(WebCore::CookieChangeEvent::CookieChangeEvent):
* Source/WebCore/Modules/cookie-store/ExtendableCookieChangeEvent.cpp:
(WebCore::ExtendableCookieChangeEvent::ExtendableCookieChangeEvent):
* Source/WebCore/Modules/encryptedmedia/MediaKeyMessageEvent.cpp:
(WebCore::MediaKeyMessageEvent::MediaKeyMessageEvent):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyMessageEvent.cpp:
(WebCore::WebKitMediaKeyMessageEvent::WebKitMediaKeyMessageEvent):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyNeededEvent.cpp:
(WebCore::WebKitMediaKeyNeededEvent::WebKitMediaKeyNeededEvent):
* Source/WebCore/Modules/gamepad/GamepadEvent.cpp:
(WebCore::GamepadEvent::GamepadEvent):
* Source/WebCore/Modules/indexeddb/IDBRequestCompletionEvent.cpp:
(WebCore::IDBRequestCompletionEvent::IDBRequestCompletionEvent):
* Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.cpp:
(WebCore::IDBVersionChangeEvent::IDBVersionChangeEvent):
* Source/WebCore/Modules/mediarecorder/BlobEvent.cpp:
(WebCore::BlobEvent::BlobEvent):
* Source/WebCore/Modules/mediarecorder/MediaRecorderErrorEvent.cpp:
(WebCore::MediaRecorderErrorEvent::MediaRecorderErrorEvent):
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.cpp:
(WebCore::BufferedChangeEvent::BufferedChangeEvent):
* Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.cpp:
(WebCore::MediaStreamTrackEvent::MediaStreamTrackEvent):
* Source/WebCore/Modules/mediastream/OverconstrainedErrorEvent.h:
* Source/WebCore/Modules/mediastream/RTCDTMFToneChangeEvent.cpp:
(WebCore::RTCDTMFToneChangeEvent::RTCDTMFToneChangeEvent):
* Source/WebCore/Modules/mediastream/RTCDataChannelEvent.cpp:
(WebCore::RTCDataChannelEvent::RTCDataChannelEvent):
* Source/WebCore/Modules/mediastream/RTCErrorEvent.cpp:
(WebCore::RTCErrorEvent::RTCErrorEvent):
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceErrorEvent.cpp:
(WebCore::RTCPeerConnectionIceErrorEvent::RTCPeerConnectionIceErrorEvent):
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceEvent.cpp:
(WebCore::RTCPeerConnectionIceEvent::RTCPeerConnectionIceEvent):
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformErrorEvent.cpp:
(WebCore::RTCRtpSFrameTransformErrorEvent::RTCRtpSFrameTransformErrorEvent):
* Source/WebCore/Modules/mediastream/RTCTrackEvent.cpp:
(WebCore::RTCTrackEvent::RTCTrackEvent):
* Source/WebCore/Modules/mediastream/RTCTransformEvent.cpp:
(WebCore::RTCTransformEvent::RTCTransformEvent):
* Source/WebCore/Modules/notifications/NotificationEvent.cpp:
(WebCore::NotificationEvent::NotificationEvent):
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
(WebCore::MerchantValidationEvent::MerchantValidationEvent):
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.cpp:
(WebCore::PaymentMethodChangeEvent::PaymentMethodChangeEvent):
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp:
(WebCore::PaymentRequestUpdateEvent::PaymentRequestUpdateEvent):
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.h:
(WebCore::PaymentRequestUpdateEvent::create):
* Source/WebCore/Modules/pictureinpicture/PictureInPictureEvent.cpp:
(WebCore::PictureInPictureEvent::PictureInPictureEvent):
* Source/WebCore/Modules/push-api/PushEvent.cpp:
(WebCore::PushEvent::PushEvent):
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp:
(WebCore::PushNotificationEvent::PushNotificationEvent):
* Source/WebCore/Modules/push-api/PushSubscriptionChangeEvent.cpp:
(WebCore::PushSubscriptionChangeEvent::PushSubscriptionChangeEvent):
* Source/WebCore/Modules/speech/SpeechRecognitionErrorEvent.cpp:
(WebCore::SpeechRecognitionErrorEvent::SpeechRecognitionErrorEvent):
* Source/WebCore/Modules/speech/SpeechRecognitionEvent.cpp:
(WebCore::SpeechRecognitionEvent::SpeechRecognitionEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisErrorEvent.cpp:
(WebCore::SpeechSynthesisErrorEvent::SpeechSynthesisErrorEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.cpp:
(WebCore::SpeechSynthesisEvent::create):
(WebCore::SpeechSynthesisEvent::SpeechSynthesisEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.h:
* Source/WebCore/Modules/webaudio/AudioProcessingEvent.cpp:
(WebCore::AudioProcessingEvent::AudioProcessingEvent):
* Source/WebCore/Modules/webaudio/OfflineAudioCompletionEvent.cpp:
(WebCore::OfflineAudioCompletionEvent::OfflineAudioCompletionEvent):
* Source/WebCore/Modules/websockets/CloseEvent.h:
* Source/WebCore/Modules/webxr/XRInputSourceEvent.cpp:
(WebCore::XRInputSourceEvent::XRInputSourceEvent):
* Source/WebCore/Modules/webxr/XRInputSourcesChangeEvent.cpp:
(WebCore::XRInputSourcesChangeEvent::XRInputSourcesChangeEvent):
* Source/WebCore/Modules/webxr/XRReferenceSpaceEvent.cpp:
(WebCore::XRReferenceSpaceEvent::XRReferenceSpaceEvent):
* Source/WebCore/Modules/webxr/XRSessionEvent.cpp:
(WebCore::XRSessionEvent::XRSessionEvent):
* Source/WebCore/animation/AnimationEventBase.cpp:
(WebCore::AnimationEventBase::AnimationEventBase):
* Source/WebCore/animation/AnimationEventBase.h:
* Source/WebCore/animation/AnimationPlaybackEvent.cpp:
(WebCore::AnimationPlaybackEvent::AnimationPlaybackEvent):
* Source/WebCore/animation/CSSAnimationEvent.cpp:
(WebCore::CSSAnimationEvent::CSSAnimationEvent):
* Source/WebCore/animation/CSSTransitionEvent.cpp:
(WebCore::CSSTransitionEvent::CSSTransitionEvent):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.cpp:
(WebCore::StyleOriginatedAnimationEvent::StyleOriginatedAnimationEvent):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.h:
* Source/WebCore/bindings/scripts/InFilesCompiler.pm:
(generateInterfacesHeader):
* Source/WebCore/css/MediaQueryListEvent.cpp:
(WebCore::MediaQueryListEvent::MediaQueryListEvent):
* Source/WebCore/dom/BeforeTextInsertedEvent.cpp:
(WebCore::BeforeTextInsertedEvent::BeforeTextInsertedEvent):
* Source/WebCore/dom/BeforeUnloadEvent.cpp:
(WebCore::BeforeUnloadEvent::BeforeUnloadEvent):
* Source/WebCore/dom/ClipboardEvent.cpp:
(WebCore::ClipboardEvent::ClipboardEvent):
* Source/WebCore/dom/CompositionEvent.cpp:
(WebCore::CompositionEvent::CompositionEvent):
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.cpp:
(WebCore::ContentVisibilityAutoStateChangeEvent::ContentVisibilityAutoStateChangeEvent):
* Source/WebCore/dom/CustomEvent.cpp:
(WebCore::CustomEvent::CustomEvent):
* Source/WebCore/dom/DeviceMotionEvent.cpp:
(WebCore::DeviceMotionEvent::DeviceMotionEvent):
* Source/WebCore/dom/DeviceOrientationEvent.cpp:
(WebCore::DeviceOrientationEvent::DeviceOrientationEvent):
* Source/WebCore/dom/DragEvent.cpp:
(WebCore::DragEvent::DragEvent):
* Source/WebCore/dom/ErrorEvent.cpp:
(WebCore::ErrorEvent::ErrorEvent):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::Event):
(WebCore::Event::create):
(WebCore::Event::createForBindings):
* Source/WebCore/dom/Event.h:
(WebCore::Event::interfaceType):
* Source/WebCore/dom/FocusEvent.cpp:
(WebCore::FocusEvent::FocusEvent):
* Source/WebCore/dom/FocusEvent.h:
* Source/WebCore/dom/FormDataEvent.cpp:
(WebCore::FormDataEvent::FormDataEvent):
* Source/WebCore/dom/HashChangeEvent.h:
* Source/WebCore/dom/InputEvent.cpp:
(WebCore::InputEvent::InputEvent):
* Source/WebCore/dom/InvokeEvent.cpp:
(WebCore::InvokeEvent::InvokeEvent):
* Source/WebCore/dom/InvokeEvent.h:
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::KeyboardEvent::KeyboardEvent):
* Source/WebCore/dom/MessageEvent.cpp:
(WebCore::MessageEvent::MessageEvent):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::create):
(WebCore::MouseEvent::createForBindings):
(WebCore::MouseEvent::MouseEvent):
* Source/WebCore/dom/MouseEvent.h:
(WebCore::MouseEvent::createForBindings): Deleted.
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::MouseRelatedEvent):
* Source/WebCore/dom/MouseRelatedEvent.h:
* Source/WebCore/dom/MutationEvent.cpp:
(WebCore::MutationEvent::MutationEvent):
* Source/WebCore/dom/MutationEvent.h:
* Source/WebCore/dom/OverflowEvent.cpp:
(WebCore::OverflowEvent::OverflowEvent):
* Source/WebCore/dom/PageTransitionEvent.cpp:
(WebCore::PageTransitionEvent::PageTransitionEvent):
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/PopStateEvent.cpp:
(WebCore::PopStateEvent::PopStateEvent):
* Source/WebCore/dom/PopStateEvent.h:
* Source/WebCore/dom/ProgressEvent.cpp:
(WebCore::ProgressEvent::ProgressEvent):
* Source/WebCore/dom/ProgressEvent.h:
(WebCore::ProgressEvent::create):
* Source/WebCore/dom/PromiseRejectionEvent.cpp:
(WebCore::PromiseRejectionEvent::PromiseRejectionEvent):
* Source/WebCore/dom/ScopedEventQueue.cpp:
(WebCore::ScopedEventQueue::dispatchEvent const):
* Source/WebCore/dom/SecurityPolicyViolationEvent.h:
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
* Source/WebCore/dom/ToggleEvent.cpp:
(WebCore::ToggleEvent::ToggleEvent):
* Source/WebCore/dom/ToggleEvent.h:
* Source/WebCore/dom/UIEvent.cpp:
(WebCore::UIEvent::UIEvent):
* Source/WebCore/dom/UIEvent.h:
(WebCore::UIEvent::create):
(WebCore::UIEvent::createForBindings):
* Source/WebCore/dom/UIEventWithKeyState.h:
(WebCore::UIEventWithKeyState::UIEventWithKeyState):
* Source/WebCore/dom/WheelEvent.cpp:
(WebCore::WheelEvent::WheelEvent):
* Source/WebCore/dom/ios/MouseEventIOS.cpp:
(WebCore::MouseEvent::create):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/make_event_factory.pl:
(generateImplementation):
* Source/WebCore/html/MediaEncryptedEvent.cpp:
(WebCore::MediaEncryptedEvent::MediaEncryptedEvent):
* Source/WebCore/html/SubmitEvent.cpp:
(WebCore::SubmitEvent::SubmitEvent):
* Source/WebCore/html/canvas/WebGLContextEvent.cpp:
(WebCore::WebGLContextEvent::WebGLContextEvent):
* Source/WebCore/html/track/TrackEvent.cpp:
(WebCore::TrackEvent::TrackEvent):
* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.cpp:
(WebCore::NavigationCurrentEntryChangeEvent::NavigationCurrentEntryChangeEvent):
* Source/WebCore/storage/StorageEvent.cpp:
(WebCore::StorageEvent::StorageEvent):
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::ExtendableEvent):
* Source/WebCore/workers/service/ExtendableEvent.h:
(WebCore::ExtendableEvent::create):
* Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
(WebCore::ExtendableMessageEvent::ExtendableMessageEvent):
* Source/WebCore/workers/service/FetchEvent.cpp:
(WebCore::FetchEvent::FetchEvent):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.cpp:
(WebCore::BackgroundFetchEvent::create):
(WebCore::BackgroundFetchEvent::BackgroundFetchEvent):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.cpp:
(WebCore::BackgroundFetchUpdateUIEvent::BackgroundFetchUpdateUIEvent):
* Source/WebCore/xml/XMLHttpRequestProgressEvent.h:
* Source/WebKitLegacy/mac/DOM/DOMEvents.mm:
(kitClass):

Canonical link: <a href="https://commits.webkit.org/274913@main">https://commits.webkit.org/274913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29e23caa07ac0dbda2a10678d659103713eb2b3a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40116 "19 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42661 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36205 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42423 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22044 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16457 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33376 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16108 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34651 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13933 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35601 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43939 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36393 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35935 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39693 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14931 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12250 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37987 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16550 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9056 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16599 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16194 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.h:
```diff
@@ -40,16 +40,16 @@ class PaymentRequestUpdateEvent : public Event {
 protected:
-    explicit PaymentRequestUpdateEvent(const AtomString& type);
-    PaymentRequestUpdateEvent(const AtomString& type, const PaymentRequestUpdateEventInit&);
+    explicit PaymentRequestUpdateEvent(enum EventInterfaceType, const AtomString& type);
```
Surprised with need `enum` here?

rniwa replied:
Yeah, it's because `EventInterface` enum has a value named `EventInterfaceType`. My plan to get rid of this enum completely once we migrate WebKitAdditions to new code.

cdumez commented on Source/WebCore/animation/StyleOriginatedAnimationEvent.h:
```diff
@@ -40,8 +40,8 @@ class StyleOriginatedAnimationEvent : public AnimationEventBase {
 protected:
-    StyleOriginatedAnimationEvent(const AtomString& type, WebAnimation*, std::optional<Seconds> scheduledTime, double, const std::optional<Style::PseudoElementIdentifier>&);
-    StyleOriginatedAnimationEvent(const AtomString&, const EventInit&, IsTrusted, double, const String&);
+    StyleOriginatedAnimationEvent(enum EventInterfaceType,const AtomString& type, WebAnimation*, std::optional<Seconds> scheduledTime, double, const std::optional<Style::PseudoElementIdentifier>&);
```
missing space after `,`

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/animation/StyleOriginatedAnimationEvent.h:
```diff
@@ -40,8 +40,8 @@ class StyleOriginatedAnimationEvent : public AnimationEventBase {
-    StyleOriginatedAnimationEvent(const AtomString& type, WebAnimation*, std::optional<Seconds> scheduledTime, double, const std::optional<Style::PseudoElementIdentifier>&);
-    StyleOriginatedAnimationEvent(const AtomString&, const EventInit&, IsTrusted, double, const String&);
+    StyleOriginatedAnimationEvent(enum EventInterfaceType,const AtomString& type, WebAnimation*, std::optional<Seconds> scheduledTime, double, const std::optional<Style::PseudoElementIdentifier>&);
+    StyleOriginatedAnimationEvent(enum EventInterfaceType,const AtomString&, const EventInit&, IsTrusted, double, const String&);
```
ditto.

rniwa replied:
Fixed

cdumez commented on Source/WebCore/dom/DeviceMotionEvent.cpp:
```diff
@@ -38,12 +38,27 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(DeviceMotionEvent);
+#else
+    // FIXME: ENABLE(DEVICE_ORIENTATION) seems to be in a strange state where
+    // it is half-guarded by #ifdefs. DeviceMotionEvent.idl is guarded
+    // but DeviceMotionEvent.cpp itself is required by ungarded code.
```
typo: `ungarded`

cdumez commented on Source/WebCore/dom/DeviceMotionEvent.cpp:
```diff
@@ -38,12 +38,27 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(DeviceMotionEvent);
+#else
+    // FIXME: ENABLE(DEVICE_ORIENTATION) seems to be in a strange state where
+    // it is half-guarded by #ifdefs. DeviceMotionEvent.idl is guarded
+    // but DeviceMotionEvent.cpp itself is required by ungarded code.
```
ditto.

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/DeviceOrientationEvent.cpp:
```diff
@@ -40,12 +40,27 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(DeviceOrientationEvent);
+#else
+    // FIXME: ENABLE(DEVICE_ORIENTATION) seems to be in a strange state where
+    // it is half-guarded by #ifdefs. DeviceOrientationEvent.idl is guarded
+    // but DeviceOrientationEvent.cpp itself is required by ungarded code.
```
ditto

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/DeviceOrientationEvent.cpp:
```diff
@@ -40,12 +40,27 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(DeviceOrientationEvent);
+#else
+    // FIXME: ENABLE(DEVICE_ORIENTATION) seems to be in a strange state where
+    // it is half-guarded by #ifdefs. DeviceOrientationEvent.idl is guarded
+    // but DeviceOrientationEvent.cpp itself is required by ungarded code.
```
ditto

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/Event.h:
```diff
@@ -72,6 +72,8 @@ class Event : public ScriptWrappable, public RefCounted<Event> {
     const AtomString& type() const { return m_type; }
     void setType(const AtomString& type) { m_type = type; }
 
+    enum EventInterfaceType interfaceType() { return m_eventInterface ? static_cast<enum EventInterfaceType>(m_eventInterface) : static_cast<enum EventInterfaceType>(eventInterface()); }
```
Should be `const`

rniwa replied:
Fixed.

cdumez reviewed (approved):
LGTM

Constellation reviewed (approved):
r=me

rniwa committed:
Devirtualize Event::eventInterface
https://bugs.webkit.org/show_bug.cgi?id=269338

Reviewed by Chris Dumez and Yusuke Suzuki.

This PR replaces Event::eventInterface, which is a virtual function, with Event::interfaceType,
which is an inline non-virtual function that reads the value off of m_eventInterface set forth
in the Event constructor. WebKitAdditions' files need to be migrated once this PR lands so they
currently fallback to calling Event::eventInterface (m_eventInterface is set to 0 in this case).

This PR also replaces EventInterface enum with EventInterfaceType enum class. Because one of
the valid values of EventInterface enum is EventInterfaceType (for Event class), we need to
explicitly write `enum EventInterfaceType` for now. The plan is to remove EventInterface once
the code in WebKitAdditions has migrated to use the new code path.

* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.cpp:
(WebCore::GPUUncapturedErrorEvent::GPUUncapturedErrorEvent):
* Source/WebCore/Modules/airplay/WebKitPlaybackTargetAvailabilityEvent.cpp:
(WebCore::WebKitPlaybackTargetAvailabilityEvent::WebKitPlaybackTargetAvailabilityEvent):
* Source/WebCore/Modules/applepay/ApplePayCancelEvent.cpp:
(WebCore::ApplePayCancelEvent::ApplePayCancelEvent):
* Source/WebCore/Modules/applepay/ApplePayCouponCodeChangedEvent.cpp:
(WebCore::ApplePayCouponCodeChangedEvent::ApplePayCouponCodeChangedEvent):
* Source/WebCore/Modules/applepay/ApplePayPaymentAuthorizedEvent.cpp:
(WebCore::ApplePayPaymentAuthorizedEvent::ApplePayPaymentAuthorizedEvent):
* Source/WebCore/Modules/applepay/ApplePayPaymentMethodSelectedEvent.cpp:
(WebCore::ApplePayPaymentMethodSelectedEvent::ApplePayPaymentMethodSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayShippingContactSelectedEvent.cpp:
(WebCore::ApplePayShippingContactSelectedEvent::ApplePayShippingContactSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayShippingMethodSelectedEvent.cpp:
(WebCore::ApplePayShippingMethodSelectedEvent::ApplePayShippingMethodSelectedEvent):
* Source/WebCore/Modules/applepay/ApplePayValidateMerchantEvent.cpp:
(WebCore::ApplePayValidateMerchantEvent::ApplePayValidateMerchantEvent):
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
(WebCore::CookieChangeEvent::CookieChangeEvent):
* Source/WebCore/Modules/cookie-store/ExtendableCookieChangeEvent.cpp:
(WebCore::ExtendableCookieChangeEvent::ExtendableCookieChangeEvent):
* Source/WebCore/Modules/encryptedmedia/MediaKeyMessageEvent.cpp:
(WebCore::MediaKeyMessageEvent::MediaKeyMessageEvent):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyMessageEvent.cpp:
(WebCore::WebKitMediaKeyMessageEvent::WebKitMediaKeyMessageEvent):
* Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeyNeededEvent.cpp:
(WebCore::WebKitMediaKeyNeededEvent::WebKitMediaKeyNeededEvent):
* Source/WebCore/Modules/gamepad/GamepadEvent.cpp:
(WebCore::GamepadEvent::GamepadEvent):
* Source/WebCore/Modules/indexeddb/IDBRequestCompletionEvent.cpp:
(WebCore::IDBRequestCompletionEvent::IDBRequestCompletionEvent):
* Source/WebCore/Modules/indexeddb/IDBVersionChangeEvent.cpp:
(WebCore::IDBVersionChangeEvent::IDBVersionChangeEvent):
* Source/WebCore/Modules/mediarecorder/BlobEvent.cpp:
(WebCore::BlobEvent::BlobEvent):
* Source/WebCore/Modules/mediarecorder/MediaRecorderErrorEvent.cpp:
(WebCore::MediaRecorderErrorEvent::MediaRecorderErrorEvent):
* Source/WebCore/Modules/mediasource/BufferedChangeEvent.cpp:
(WebCore::BufferedChangeEvent::BufferedChangeEvent):
* Source/WebCore/Modules/mediastream/MediaStreamTrackEvent.cpp:
(WebCore::MediaStreamTrackEvent::MediaStreamTrackEvent):
* Source/WebCore/Modules/mediastream/OverconstrainedErrorEvent.h:
* Source/WebCore/Modules/mediastream/RTCDTMFToneChangeEvent.cpp:
(WebCore::RTCDTMFToneChangeEvent::RTCDTMFToneChangeEvent):
* Source/WebCore/Modules/mediastream/RTCDataChannelEvent.cpp:
(WebCore::RTCDataChannelEvent::RTCDataChannelEvent):
* Source/WebCore/Modules/mediastream/RTCErrorEvent.cpp:
(WebCore::RTCErrorEvent::RTCErrorEvent):
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceErrorEvent.cpp:
(WebCore::RTCPeerConnectionIceErrorEvent::RTCPeerConnectionIceErrorEvent):
* Source/WebCore/Modules/mediastream/RTCPeerConnectionIceEvent.cpp:
(WebCore::RTCPeerConnectionIceEvent::RTCPeerConnectionIceEvent):
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransformErrorEvent.cpp:
(WebCore::RTCRtpSFrameTransformErrorEvent::RTCRtpSFrameTransformErrorEvent):
* Source/WebCore/Modules/mediastream/RTCTrackEvent.cpp:
(WebCore::RTCTrackEvent::RTCTrackEvent):
* Source/WebCore/Modules/mediastream/RTCTransformEvent.cpp:
(WebCore::RTCTransformEvent::RTCTransformEvent):
* Source/WebCore/Modules/notifications/NotificationEvent.cpp:
(WebCore::NotificationEvent::NotificationEvent):
* Source/WebCore/Modules/paymentrequest/MerchantValidationEvent.cpp:
(WebCore::MerchantValidationEvent::MerchantValidationEvent):
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.cpp:
(WebCore::PaymentMethodChangeEvent::PaymentMethodChangeEvent):
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.cpp:
(WebCore::PaymentRequestUpdateEvent::PaymentRequestUpdateEvent):
* Source/WebCore/Modules/paymentrequest/PaymentRequestUpdateEvent.h:
(WebCore::PaymentRequestUpdateEvent::create):
* Source/WebCore/Modules/pictureinpicture/PictureInPictureEvent.cpp:
(WebCore::PictureInPictureEvent::PictureInPictureEvent):
* Source/WebCore/Modules/push-api/PushEvent.cpp:
(WebCore::PushEvent::PushEvent):
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp:
(WebCore::PushNotificationEvent::PushNotificationEvent):
* Source/WebCore/Modules/push-api/PushSubscriptionChangeEvent.cpp:
(WebCore::PushSubscriptionChangeEvent::PushSubscriptionChangeEvent):
* Source/WebCore/Modules/speech/SpeechRecognitionErrorEvent.cpp:
(WebCore::SpeechRecognitionErrorEvent::SpeechRecognitionErrorEvent):
* Source/WebCore/Modules/speech/SpeechRecognitionEvent.cpp:
(WebCore::SpeechRecognitionEvent::SpeechRecognitionEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisErrorEvent.cpp:
(WebCore::SpeechSynthesisErrorEvent::SpeechSynthesisErrorEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.cpp:
(WebCore::SpeechSynthesisEvent::create):
(WebCore::SpeechSynthesisEvent::SpeechSynthesisEvent):
* Source/WebCore/Modules/speech/SpeechSynthesisEvent.h:
* Source/WebCore/Modules/webaudio/AudioProcessingEvent.cpp:
(WebCore::AudioProcessingEvent::AudioProcessingEvent):
* Source/WebCore/Modules/webaudio/OfflineAudioCompletionEvent.cpp:
(WebCore::OfflineAudioCompletionEvent::OfflineAudioCompletionEvent):
* Source/WebCore/Modules/websockets/CloseEvent.h:
* Source/WebCore/Modules/webxr/XRInputSourceEvent.cpp:
(WebCore::XRInputSourceEvent::XRInputSourceEvent):
* Source/WebCore/Modules/webxr/XRInputSourcesChangeEvent.cpp:
(WebCore::XRInputSourcesChangeEvent::XRInputSourcesChangeEvent):
* Source/WebCore/Modules/webxr/XRReferenceSpaceEvent.cpp:
(WebCore::XRReferenceSpaceEvent::XRReferenceSpaceEvent):
* Source/WebCore/Modules/webxr/XRSessionEvent.cpp:
(WebCore::XRSessionEvent::XRSessionEvent):
* Source/WebCore/animation/AnimationEventBase.cpp:
(WebCore::AnimationEventBase::AnimationEventBase):
* Source/WebCore/animation/AnimationEventBase.h:
* Source/WebCore/animation/AnimationPlaybackEvent.cpp:
(WebCore::AnimationPlaybackEvent::AnimationPlaybackEvent):
* Source/WebCore/animation/CSSAnimationEvent.cpp:
(WebCore::CSSAnimationEvent::CSSAnimationEvent):
* Source/WebCore/animation/CSSTransitionEvent.cpp:
(WebCore::CSSTransitionEvent::CSSTransitionEvent):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.cpp:
(WebCore::StyleOriginatedAnimationEvent::StyleOriginatedAnimationEvent):
* Source/WebCore/animation/StyleOriginatedAnimationEvent.h:
* Source/WebCore/bindings/scripts/InFilesCompiler.pm:
(generateInterfacesHeader):
* Source/WebCore/css/MediaQueryListEvent.cpp:
(WebCore::MediaQueryListEvent::MediaQueryListEvent):
* Source/WebCore/dom/BeforeTextInsertedEvent.cpp:
(WebCore::BeforeTextInsertedEvent::BeforeTextInsertedEvent):
* Source/WebCore/dom/BeforeUnloadEvent.cpp:
(WebCore::BeforeUnloadEvent::BeforeUnloadEvent):
* Source/WebCore/dom/ClipboardEvent.cpp:
(WebCore::ClipboardEvent::ClipboardEvent):
* Source/WebCore/dom/CompositionEvent.cpp:
(WebCore::CompositionEvent::CompositionEvent):
* Source/WebCore/dom/ContentVisibilityAutoStateChangeEvent.cpp:
(WebCore::ContentVisibilityAutoStateChangeEvent::ContentVisibilityAutoStateChangeEvent):
* Source/WebCore/dom/CustomEvent.cpp:
(WebCore::CustomEvent::CustomEvent):
* Source/WebCore/dom/DeviceMotionEvent.cpp:
(WebCore::DeviceMotionEvent::DeviceMotionEvent):
* Source/WebCore/dom/DeviceOrientationEvent.cpp:
(WebCore::DeviceOrientationEvent::DeviceOrientationEvent):
* Source/WebCore/dom/DragEvent.cpp:
(WebCore::DragEvent::DragEvent):
* Source/WebCore/dom/ErrorEvent.cpp:
(WebCore::ErrorEvent::ErrorEvent):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::Event):
(WebCore::Event::create):
(WebCore::Event::createForBindings):
* Source/WebCore/dom/Event.h:
(WebCore::Event::interfaceType):
* Source/WebCore/dom/FocusEvent.cpp:
(WebCore::FocusEvent::FocusEvent):
* Source/WebCore/dom/FocusEvent.h:
* Source/WebCore/dom/FormDataEvent.cpp:
(WebCore::FormDataEvent::FormDataEvent):
* Source/WebCore/dom/HashChangeEvent.h:
* Source/WebCore/dom/InputEvent.cpp:
(WebCore::InputEvent::InputEvent):
* Source/WebCore/dom/InvokeEvent.cpp:
(WebCore::InvokeEvent::InvokeEvent):
* Source/WebCore/dom/InvokeEvent.h:
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::KeyboardEvent::KeyboardEvent):
* Source/WebCore/dom/MessageEvent.cpp:
(WebCore::MessageEvent::MessageEvent):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::create):
(WebCore::MouseEvent::createForBindings):
(WebCore::MouseEvent::MouseEvent):
* Source/WebCore/dom/MouseEvent.h:
(WebCore::MouseEvent::createForBindings): Deleted.
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::MouseRelatedEvent):
* Source/WebCore/dom/MouseRelatedEvent.h:
* Source/WebCore/dom/MutationEvent.cpp:
(WebCore::MutationEvent::MutationEvent):
* Source/WebCore/dom/MutationEvent.h:
* Source/WebCore/dom/OverflowEvent.cpp:
(WebCore::OverflowEvent::OverflowEvent):
* Source/WebCore/dom/PageTransitionEvent.cpp:
(WebCore::PageTransitionEvent::PageTransitionEvent):
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/PopStateEvent.cpp:
(WebCore::PopStateEvent::PopStateEvent):
* Source/WebCore/dom/PopStateEvent.h:
* Source/WebCore/dom/ProgressEvent.cpp:
(WebCore::ProgressEvent::ProgressEvent):
* Source/WebCore/dom/ProgressEvent.h:
(WebCore::ProgressEvent::create):
* Source/WebCore/dom/PromiseRejectionEvent.cpp:
(WebCore::PromiseRejectionEvent::PromiseRejectionEvent):
* Source/WebCore/dom/ScopedEventQueue.cpp:
(WebCore::ScopedEventQueue::dispatchEvent const):
* Source/WebCore/dom/SecurityPolicyViolationEvent.h:
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
* Source/WebCore/dom/ToggleEvent.cpp:
(WebCore::ToggleEvent::ToggleEvent):
* Source/WebCore/dom/ToggleEvent.h:
* Source/WebCore/dom/UIEvent.cpp:
(WebCore::UIEvent::UIEvent):
* Source/WebCore/dom/UIEvent.h:
(WebCore::UIEvent::create):
(WebCore::UIEvent::createForBindings):
* Source/WebCore/dom/UIEventWithKeyState.h:
(WebCore::UIEventWithKeyState::UIEventWithKeyState):
* Source/WebCore/dom/WheelEvent.cpp:
(WebCore::WheelEvent::WheelEvent):
* Source/WebCore/dom/ios/MouseEventIOS.cpp:
(WebCore::MouseEvent::create):
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/make_event_factory.pl:
(generateImplementation):
* Source/WebCore/html/MediaEncryptedEvent.cpp:
(WebCore::MediaEncryptedEvent::MediaEncryptedEvent):
* Source/WebCore/html/SubmitEvent.cpp:
(WebCore::SubmitEvent::SubmitEvent):
* Source/WebCore/html/canvas/WebGLContextEvent.cpp:
(WebCore::WebGLContextEvent::WebGLContextEvent):
* Source/WebCore/html/track/TrackEvent.cpp:
(WebCore::TrackEvent::TrackEvent):
* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
* Source/WebCore/page/NavigationCurrentEntryChangeEvent.cpp:
(WebCore::NavigationCurrentEntryChangeEvent::NavigationCurrentEntryChangeEvent):
* Source/WebCore/storage/StorageEvent.cpp:
(WebCore::StorageEvent::StorageEvent):
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::ExtendableEvent):
* Source/WebCore/workers/service/ExtendableEvent.h:
(WebCore::ExtendableEvent::create):
* Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
(WebCore::ExtendableMessageEvent::ExtendableMessageEvent):
* Source/WebCore/workers/service/FetchEvent.cpp:
(WebCore::FetchEvent::FetchEvent):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.cpp:
(WebCore::BackgroundFetchEvent::create):
(WebCore::BackgroundFetchEvent::BackgroundFetchEvent):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.cpp:
(WebCore::BackgroundFetchUpdateUIEvent::BackgroundFetchUpdateUIEvent):
* Source/WebCore/xml/XMLHttpRequestProgressEvent.h:
* Source/WebKitLegacy/mac/DOM/DOMEvents.mm:
(kitClass):

Canonical link: https://commits.webkit.org/274913@main
